nixos/modules/services/video/wivrn.nix

```diff
@@ -0,0 +1,137 @@
+{ config, pkgs, lib, ... }:
+let
+  inherit (lib) mkAliasOptionModule mkIf mkEnableOption mkPackageOption mkOption mkDefault optionalString getExe' maintainers;
+  cfg = config.services.wivrn;
+  configFormat = pkgs.formats.json { };
+  configFile = configFormat.generate "config.json" cfg.config.json;
+in
+{
+  options = {
+    services.wivrn = {
+      enable = mkEnableOption "WiVRn, an OpenXR streaming application";
+
+      package = mkPackageOption pkgs "wivrn" { };
+
+      openFirewall = mkEnableOption "the default ports in the firewall for the WiVRn server";
+
+      defaultRuntime = mkEnableOption ''
+        WiVRn Monado as the default OpenXR runtime on the system. The config can be found at `/etc/xdg/openxr/1/active_runtime.json`.
+
+        Note that applications can bypass this option by setting an active
+        runtime in a writable XDG_CONFIG_DIRS location like `~/.config`
+      '' // { default = true; };
+
+      highPriority = mkEnableOption "high priority capability for wivrn-server" // { default = true; };
+
+      config = {
+        enable = mkEnableOption "configuration for WiVRn";
+        json = mkOption {
+          type = configFormat.type;
+          description = ''
+            Configuration for WiVRn. The attributes are serialized to JSON in config.json.
+            See https://github.com/Meumeu/WiVRn/blob/master/docs/configuration.md
+          '';
+          example = {
+            scale = 1.0;
+            bitrate = 100000000;
+            encoders = [
+              {
+                encoder = "nvenc";
+                codec = "h264";
+                width = 0.5;
+                height = 1.0;
+                offset_x = 0.0;
+                offset_y = 0.0;
+              }
+              {
+                encoder = "nvenc";
+                codec = "h264";
+                width = 0.5;
+                height = 1.0;
+                offset_x = 0.5;
+                offset_y = 0.0;
+              }
+            ];
+          };
+        };
+      };
+    };
+  };
+
+  imports = [
+    (mkAliasOptionModule [ "services" "wivrn" "monadoEnvironment" ] [ "systemd" "user" "services" "wivrn" "environment" ])
+  ];
+
+  config = mkIf cfg.enable {
+    security.wrappers."wivrn-server" = mkIf cfg.highPriority {
+      setuid = false;
+      owner = "root";
+      group = "root";
+      # cap_sys_nice needed for asynchronous reprojection
+      capabilities = "cap_sys_nice+eip";
+      source = getExe' cfg.package "wivrn-server";
+    };
+
+    systemd.user = {
+      services.wivrn = {
+        description = "WiVRn XR runtime service module";
```

Comment:
Nit: Is there any hardening that can be done here?

Reply:
Personally I don't know. Xytovl can probably provide some input since they develop WiVRn, and maybe Scrumplex too since they know more about Monado than I do.

Reply:
It probably makes sense to set `ProtectSystem`, `InaccessiblePaths`, `ReadWritePaths`, and `ReadOnlyPaths`. For capabilities, use `CapabilityBoundingSet` and `AmbientCapabilities`.

See https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html

Reply:
The unit file is shipped upstream: https://github.com/Meumeu/WiVRn/blob/4c9f24499eb9bbea3c8949e026b6177e19e7c1a3/server/dist/wivrn.service.in
There is no sandboxing yet, I suggest raising an issue on https://github.com/Meumeu/WiVRn if you want it.

Reply:
Nixpkgs tries to minimally deviate from upstream. NixOS *doesn't*, and hardening options are actually explicitly called out as one of the things that are frequently adjusted: https://discourse.nixos.org/t/re-tensions-between-upstream-developers-and-package-maintainers/19556/6?u=pandapip1

Reply:
How secure do we want the service to be? I did a bit of reading and got the security from like a 9.8 down to a 5 in systemd-analyze security, all while still being functional.

Reply:
The goal is to provide the minimum attack surface, but even a 5 is *much* better compared to a 9.8. I would consider that more than enough hardening for a new nixpkgs module.